contracts/zap/staking/SolidZapStaker.sol

```diff
@@ -1,10 +1,170 @@
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "../../interfaces/staking/ISolidZapStaker.sol";
+import "../../interfaces/staking/ISolidStakingActions_0_8_18.sol";
+import "../../interfaces/liquidity-deployer/IHypervisor_0_8_18.sol";
+import "../../interfaces/liquidity-deployer/IUniProxy_0_8_18.sol";
+import "../../libraries/GPv2SafeERC20_0_8_18.sol";
+
+/// @author Solid World
+contract SolidZapStaker is ISolidZapStaker, ReentrancyGuard {
+    using GPv2SafeERC20 for IERC20;
+
     address public immutable router;
+    address public immutable iUniProxy;
+    address public immutable solidStaking;
 
-    constructor(address _router) {
+    struct HypervisorTokens {
+        address token0;
+        address token1;
+    }
+
+    struct TokenBalances {
+        uint token0Balance;
+        uint token1Balance;
+    }
+
+    constructor(
+        address _router,
+        address _iUniProxy,
+        address _solidStaking
+    ) {
         router = _router;
+        iUniProxy = _iUniProxy;
+        solidStaking = _solidStaking;
+    }
+
+    /// @inheritdoc ISolidZapStaker
+    function stakeDoubleSwap(
+        address inputToken,
+        uint inputAmount,
+        address hypervisor,
+        bytes calldata swap1,
+        bytes calldata swap2,
+        uint minShares,
+        address recipient
+    ) external nonReentrant returns (uint) {
+        return _stakeDoubleSwap(inputToken, inputAmount, hypervisor, swap1, swap2, minShares, recipient);
+    }
+
+    /// @inheritdoc ISolidZapStaker
+    function stakeDoubleSwap(
+        address inputToken,
+        uint inputAmount,
+        address hypervisor,
+        bytes calldata swap1,
+        bytes calldata swap2,
+        uint minShares
+    ) external nonReentrant returns (uint) {
+        return _stakeDoubleSwap(inputToken, inputAmount, hypervisor, swap1, swap2, minShares, msg.sender);
+    }
+
+    function _stakeDoubleSwap(
+        address inputToken,
+        uint inputAmount,
+        address hypervisor,
+        bytes calldata swap1,
+        bytes calldata swap2,
+        uint minShares,
+        address recipient
+    ) private returns (uint) {
+        IERC20(inputToken).safeTransferFrom(msg.sender, address(this), inputAmount);
+        _approveTokenSpendingIfNeeded(inputToken, router);
+
+        HypervisorTokens memory tokens = _fetchHypervisorTokens(hypervisor);
+        TokenBalances memory balancesBeforeSwap = _fetchTokenBalances(tokens);
+        _swapViaRouter(swap1);
+        _swapViaRouter(swap2);
+        TokenBalances memory balancesAfterSwap = _fetchTokenBalances(tokens);
+
+        _approveTokenSpendingIfNeeded(tokens.token0, hypervisor);
+        _approveTokenSpendingIfNeeded(tokens.token1, hypervisor);
+        uint shares = _deployLiquidity(
+            balancesAfterSwap.token0Balance - balancesBeforeSwap.token0Balance,
+            balancesAfterSwap.token1Balance - balancesBeforeSwap.token1Balance,
```

Comment:
Could you explain why deploy the difference? I thought we need to deploy everything that was swapped?

Reply:
Yes, it's meant to protect against a denial of service attack. Because before doing the stake, we run some simulations and for example we expect to swap 50 USDC and 2 CRISP-M, which can be deposited into Gamma. If someone sends 5 USDC to the contract before we do the stake, we'll end up with a balance of 55 USDC and 2 CRISP-M, which makes the deposit into Gamma revert

To prevent this we work with balances before and after swaps

Reply:
Thanks for explanation, now I see